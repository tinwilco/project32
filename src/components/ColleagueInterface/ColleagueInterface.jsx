import { useMemo, useState } from "react";
import { format } from "date-fns";
import { Alert, Button, Spinner } from "reactstrap";
import PropTypes from "prop-types";

import EmojiButton from "../EmojiButton/EmojiButton";
import UsernameEntry from "../UsernameEntry/UsernameEntry";

import styles from "./ColleagueInterface.module.css";
import lightstyles from "../LightApp.module.css";

import { buttons as emojiButtonArray } from "../../assets/emojiButtons.json";
import { SendRecord } from "../utils/api";

const ColleagueInterface = ({ darkMode }) => {
  const selectedDate = useMemo(() => new Date(), []);

  const [selectedMood, updateSelectedMood] = useState(null);
  const [userName, updateUserName] = useState("");
  const [isLoading, updateIsloading] = useState(false);
  const [isSubmitted, updateIsSubmitted] = useState(false);
  const [isError, updateIsError] = useState(false);

  const handleSubmit = async () => {
    updateIsloading(true);

    const payload = {
      userName,
      mood: selectedMood,
      moodDate: format(selectedDate, "yyyy-MM-dd"),
      comment: "",
    };

    const response = await SendRecord(payload);

    if (response.status === 200) {
      updateIsloading(false);
      updateIsSubmitted(true);
      updateIsError(false);
    } else {
      updateIsloading(false);
      updateIsSubmitted(false);
      updateIsError(true);
    }
  };

  return (
    <>
      <h2
        className={
          darkMode ? styles.TextBackground : lightstyles.TextBackground
        }
      >
        {`Hello ${userName}, How are you feeling today?`}
      </h2>

      <h3
        className={
          darkMode ? styles.TextBackground : lightstyles.TextBackground
        }
      >
        {format(selectedDate, "EEEE dd/MM/yyyy")}
      </h3>
      {!selectedMood && (
        <p
          className={
            darkMode ? styles.HelpBackground : lightstyles.HelpBackground
          }
        >
          Please select from the options below
        </p>
      )}
      <div className={styles.App_Button_Container}>
        {emojiButtonArray.map((button) => (
          <EmojiButton
            label={button.label}
            emoji={button.emoji}
            key={`emoji-button-${button.label}`}
            handleButtonPress={() => updateSelectedMood(button.label)}
            selected={selectedMood === button.label}
          />
        ))}
      </div>
      <UsernameEntry
        handleUpdateUsername={(value) => updateUserName(value)}
        username={userName}
        darkMode={darkMode}
      />
      <Button
        color="primary"
        size="lg"
        className={styles.submitButton}
        onClick={handleSubmit}
        disabled={isLoading || isSubmitted || !userName || !selectedMood}
      >
        Submit
        {isLoading && (
          <Spinner className={styles.spinner} size="sm" color="light" />
        )}
      </Button>
      {isSubmitted && (
        <Alert color="success" className={styles.alert}>
          Thanks. Your mood has been recorded.
        </Alert>
      )}
      {isError && (
        <Alert color="danger" className={styles.alert}>
          Sorry, your response could not be recorded at this time.
        </Alert>
      )}
    </>
  );
};

ColleagueInterface.propTypes = {
  darkMode: PropTypes.string.isRequired,
};

ColleagueInterface.defaultProps = {
  darkMode: "",
};

export default ColleagueInterface;
