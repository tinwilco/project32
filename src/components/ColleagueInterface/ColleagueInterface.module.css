@import "../../assets/colors.css";
@import "../../assets/breakpoints.css";

.App_Button_Container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
}

.submitButton {
  margin-top: 20px;
  display: flex;
  align-items: center;
}

.helpText {
  font-size: 16px;
  font-style: italic;
}

.spinner {
  margin-left: 10px;
}

.alert {
  font-size: 16px;
  padding: 5px;
  margin: 20px;
  max-width: 400px;
}

.TextBackground {
  transition-duration: 2s;
  color: whte;
}

.HelpText {
  font-size: 16px;
  transition-duration: 2s;
  color: white;
}
