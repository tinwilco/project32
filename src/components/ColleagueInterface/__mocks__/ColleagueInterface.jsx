const B = () => {
  return <p>ColleagueInterface</p>;
};

export default B;
