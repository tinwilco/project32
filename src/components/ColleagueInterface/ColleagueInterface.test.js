import React from "react";
import { render, screen } from "@testing-library/react";
import MockDate from "mockdate";

import ColleagueInterface from "./ColleagueInterface";

describe("Colleague Interface component", () => {
  beforeAll(() => {
    MockDate.set("2020-01-01");
  });
  afterAll(() => {
    MockDate.reset();
  });

  it("matches the snapshot", () => {
    const subject = render(<ColleagueInterface />);
    expect(subject.baseElement).toMatchSnapshot();
  });

  it("displays a success alert when submitted", () => {
    render(<ColleagueInterface />);
    expect(screen.getByText("Submit").closest("button")).toBeDisabled();
  });
});
