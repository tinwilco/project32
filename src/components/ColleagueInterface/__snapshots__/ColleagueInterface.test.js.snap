// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colleague Interface component matches the snapshot 1`] = `
<body>
  <div>
    <h2
      class="TextBackground"
    >
      Hello , How are you feeling today?
    </h2>
    <h3
      class="TextBackground"
    >
      Wednesday 01/01/2020
    </h3>
    <p
      class="HelpText"
    >
      Please select from the options below
    </p>
    <div
      class="App_Button_Container"
    >
      <button
        class="EmojiButton"
        type="button"
      >
        <span
          aria-label="miserable"
          role="img"
        >
          😢
        </span>
      </button>
      <button
        class="EmojiButton"
        type="button"
      >
        <span
          aria-label="sad"
          role="img"
        >
          🙁
        </span>
      </button>
      <button
        class="EmojiButton"
        type="button"
      >
        <span
          aria-label="neutral"
          role="img"
        >
          😑
        </span>
      </button>
      <button
        class="EmojiButton"
        type="button"
      >
        <span
          aria-label="happy"
          role="img"
        >
          🙂
        </span>
      </button>
      <button
        class="EmojiButton"
        type="button"
      >
        <span
          aria-label="ecstatic"
          role="img"
        >
          🥳
        </span>
      </button>
    </div>
    <label
      class="TextBackground"
      for="usernameInputTextBox"
    >
      Who are you?
    </label>
    <input
      class="UsernameEntry form-control-lg form-control"
      id="usernameInputTextBox"
      placeholder="Enter your name"
      type="text"
      value=""
    />
    <button
      class="submitButton btn btn-primary btn-lg disabled"
      disabled=""
      type="button"
    >
      Submit
    </button>
  </div>
</body>
`;
